### ISSUE TITLE
ArgumentOutOfRangeException: Completion of Splat, but Not $

### ISSUE THREAD

**Issue by msftrncs** [PowerShell/PSReadLine#983]
Using a custom build of PS 7 Preview, but also applies to 5.1 and 6/6.1/6.2 in the right conditions.

I have been working with the PowerShell code to try to get completion starting at a splat `@` to work.  With that working, the current completion code completes the `$` variable as `@{$}`, which is knowingly invalid, and it causes an ArgumentOutOfRange exception.

## Steps to reproduce or exception report

This reproduction is for the released builds of PowerShell, as they will not currently produce completions for just a `@` argument.

Complete the following:
```PowerShell
echo @?<CTRL SPACE> # crashes after list is displayed, but before the first result is inserted
echo $<CTRL SPACE> # notice selection varies with variable names that require a brace
```

Note, the window must be large enough as to be able to display the entire list without the input line scrolling off the top.  Also, I've only tested this in Windows.

## Code in Question

I traced down the code that is calling the exception:
https://github.com/PowerShell/PSReadLine/blob/b7b8f7bb1dd094cc86fc71ea74e7e8427b6415e0/PSReadLine/Completion.cs#L732-L737
I think the check for a `{` is incorrect here, and that there is no reason to ever substring the original text, removing the prefix, which causes the completion to look like its replacing the prefix as well.

**Comment by msftrncs** [comment]
I'm also pretty sure there is some code somewhere else dealing specifically with the `@` condition, and its not just the linked to function above that's at problem, because, why else would it exist.

**Comment by msftrncs** [comment]
Also reported at https://github.com/PowerShell/PowerShell/issues/10243

**Comment by msftrncs** [comment]
I think the error is here:

https://github.com/PowerShell/PSReadLine/blob/b7b8f7bb1dd094cc86fc71ea74e7e8427b6415e0/PSReadLine/Completion.cs#L102-L114

The `CompletionResultType` is `Variable` for a splat, and so PSReadLine for some reason is tearing off the braces, but while doing so changes the `@` to a `$`, and then later attempts to compare that with each completion to see how much of the command its completing for highlighting purposes, but since the `@` is now a `$`, the comparison determines that no text changes (or stays the same or whatever), which results in an empty string which then ends up as the `userCompletionText` in `FindUserCompletionTextPosition` which because the second character of the `match.CompletionText` is a '{' results in a substring(1) of an empty string, ArgumentOutOfRangeException.

**Comment by daxian-dbw** [comment]
> echo @?<CTRL SPACE> # crashes after list is displayed, but before the first result is inserted
echo $<CTRL SPACE> # notice selection varies with variable names that require a brace

@msftrncs I cannot repro the crash with these steps. I'm using `7.0-preview.1` (see the screen gif below).
Does the console buffer size need to be somehow configured to repro this?

![crash-not-repro](https://user-images.githubusercontent.com/127450/62190125-733c3c00-b325-11e9-9b20-bffb6a2441c8.gif)
![crash-not-repro2](https://user-images.githubusercontent.com/127450/62190135-76372c80-b325-11e9-903e-2e3fd6f93581.gif)

**Comment by msftrncs** [comment]
@daxian-dbw, yes, the buffer needs enough lines visible so that the list can populate without the input line scrolling off the top, so that the first item in the list can replace the current input.

 
![image](https://user-images.githubusercontent.com/26179051/62234271-f98d6800-b38f-11e9-85df-bf1220151e6a.png)

**Comment by daxian-dbw** [comment]
> yes, the buffer needs enough lines visible so that the list can populate without the input line scrolling off the top, so that the first item in the list can replace the current input.

Thanks, that helps. I can repro the issue now.

**Comment by msftrncs** [comment]
I think I forgot to include that `echo @?<TAB>` is also a quick way to produce the error, as long as the first completion returned is the typical `@{$}`.

oops, guess it doesn't, that works...

**Comment by msftrncs** [comment]
I think I figured out what the check for a `{` is doing …

The detection for a `{` appears to be to handle the situation where `@en` is the input, and then when `@{env:blah(blah)}` is the completion, the selection becomes 'v:blah(blah)}`, preventing the matching process from selecting from the first brace.

**Comment by daxian-dbw** [comment]
@msftrncs I can reproduce the crash using `echo @?<CTRL SPACE>`, but not exactly sure what you mean with the second repro `echo $<CTRL SPACE>`.
> \# notice selection varies with variable names that require a brace

The selection looks fine to me for variables that require a brace. Can you please clarify it a bit more?

![image](https://user-images.githubusercontent.com/127450/64392041-e5c4c980-cfff-11e9-8888-0f2ad7261d15.png)

**Comment by msftrncs** [comment]
> >     echo $<CTRL SPACE> # notice selection varies with variable names that require a brace
> The selection looks fine to me for variables that require a brace. Can you please clarify it a bit more?

Hopefully these snips clarify the comment, but because it could be said better: 'the selection of the leading `$` varies depending on whether the variable name requires a brace, or not.' 

![image](https://user-images.githubusercontent.com/26179051/64397160-d6587700-d025-11e9-8974-7e8a740f222b.png)
![image](https://user-images.githubusercontent.com/26179051/64397223-26373e00-d026-11e9-96f5-373488833b06.png)
![image](https://user-images.githubusercontent.com/26179051/64397262-4ebf3800-d026-11e9-8034-7fe5a523e861.png)

Because the `$` sigil is there in all cases, it should not be part of the selection.  The fix in PR #984 acts like this:
![image](https://user-images.githubusercontent.com/26179051/64397889-74e5d780-d028-11e9-858e-e74834d85da1.png)
![image](https://user-images.githubusercontent.com/26179051/64397929-8dee8880-d028-11e9-9af2-9137f9c9370b.png)
![image](https://user-images.githubusercontent.com/26179051/64397942-a068c200-d028-11e9-9258-d113d1ea0a6c.png)

**Comment by daxian-dbw** [comment]
Thanks for the further clarification. You really have a pair of sharp eyes to find this inconsistency 😄
